### PR TITLE
RSDK-2253 ignore StreamStatus Unimplemented errors

### PR DESCRIPTION
**Pull request by bazile-clyde**
Tested locally with the [Intel® RealSense™](https://www.intelrealsense.com/) camera. This removes the `Unimplemented` errors but there are still `WARN`ings (see below). Let me know if I should remove those too.

```bash
pi@raspberrypi:~/rdk $ go run web/cmd/server/main.go -config /etc/viam.json 
...
2023-03-27T17:26:18.969-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-27T17:26:18-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "SendSessionHeartbeat", "grpc.code": "OK", "grpc.time_ms": 0.279}

2023-03-27T17:26:19.215-0400    WARN    robot_server    zap/options.go:212      finished client unary call      {"system": "grpc", "span.kind": "client", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetStatus", "error": "rpc error: code = Unimplemented desc = ", "grpc.code": "Unimplemented", "grpc.time_ms": 0.914}

2023-03-27T17:26:19.368-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-27T17:26:19-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "ResourceNames", "grpc.code": "OK", "grpc.time_ms": 0.185}
2023-03-27T17:26:19.377-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-27T17:26:19-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetOperations", "grpc.code": "OK", "grpc.time_ms": 0.219}
2023-03-27T17:26:19.379-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-27T17:26:19-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "SendSessionHeartbeat", "grpc.code": "OK", "grpc.time_ms": 0.153}
2023-03-27T17:26:19.392-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-27T17:26:19-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetSessions", "grpc.code": "OK", "grpc.time_ms": 0.212}
2023-03-27T17:26:19.714-0400    WARN    robot_server    zap/options.go:212      finished client unary call      {"system": "grpc", "span.kind": "client", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetStatus", "error": "rpc error: code = Unimplemented desc = ", "grpc.code": "Unimplemented", "grpc.time_ms": 0.838}
2023-03-27T17:26:19.809-0400  INFO    robot_server    zap/options.go:212      finished streaming call with code Canceled      {"grpc.start_time": "2023-03-27T17:26:04-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "proto.rpc.webrtc.v1.SignalingService", "grpc.method": "Answer", "error": "rpc error: code = Canceled desc = context canceled", "grpc.code": "Canceled", "grpc.time_ms": 14919.015}
...
```

### CODE REVIEW

File: robot/server/server.go [hunk 1]
```diff
@@ -322,20 +325,19 @@ func (s *Server) StreamStatus(req *pb.StreamStatusRequest, streamServer pb.Robot
...
-		select {
-		case <-streamServer.Context().Done():
+		if !utils.SelectContextOrWaitChan(streamServer.Context(), ticker.C) {
 			return streamServer.Context().Err()
-		default:
-		}
-		select {
-		case <-streamServer.Context().Done():
-			return streamServer.Context().Err()
-		case <-ticker.C:
```
Comment (bazile-clyde) [inline robot/server/server.go:333]: Refactored to use the `SelectContextOrWaitChan` function which does the exact the same thing.

**Comment by bhaney** [comment]
Do the warnings overwhelm the viam-server logs in the same way as the errors?

File: robot/server/server.go [hunk 2]
```diff
@@ -322,20 +324,19 @@ func (s *Server) StreamStatus(req *pb.StreamStatusRequest, streamServer pb.Robot
...
 		}
+
 		status, err := s.GetStatus(streamServer.Context(), &pb.GetStatusRequest{ResourceNames: req.ResourceNames})
-		if err != nil {
+		switch {
+		case err == nil:
+		case grpcstatus.Code(err) == codes.Unimplemented:
+			return nil
```
Comment (bazile-clyde) [inline robot/server/server.go:335]: If it's Unimplemented there's no sense in staying in this for loop, repeatedly making calls to `GetStatus` every tick.

**Review by bhaney — APPROVED** [review]
LGTM! I think the only thing to check is if the WARN messages are also just as numerous in the logs. But with your exit from the for loop, I think it will prevent the logs from being spammed.

**Comment by bazile-clyde** [comment]
> LGTM! I think the only thing to check is if the WARN messages are also just as numerous in the logs. But with your exit from the for loop, I think it will prevent the logs from being spammed.

That lessened the logs but it's still a lot. It seems something, probably on the front-end, is making repeated requests to this function. I think changing that would be out of scope for this ticket but can file a new ticket to address it if you think it's worthwhile.

Here's a sample for how often it occurs now (see the `WARN`s at the top and bottom)
```bash
2023-03-30T15:48:06.038-0400    WARN    robot_server    zap/options.go:212      finished client unary call      {"system": "grpc", "span.kind": "client", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetStatus", "error": "rpc error: code = Unimplemented desc = ", "grpc.code": "Unimplemented", "grpc.time_ms": 1.059}
2023-03-30T15:48:06.038-0400    INFO    robot_server    zap/options.go:212      finished streaming call with code OK    {"grpc.start_time": "2023-03-30T15:48:05-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "StreamStatus", "grpc.code": "OK", "grpc.time_ms": 502.436}
2023-03-30T15:48:06.040-0400    WARN    robot_server    zap/options.go:212      finished client unary call      {"system": "grpc", "span.kind": "client", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetStatus", "error": "rpc error: code = Unimplemented desc = ", "grpc.code": "Unimplemented", "grpc.time_ms": 0.972}
2023-03-30T15:48:06.040-0400    INFO    robot_server    zap/options.go:212      finished streaming call with code OK    {"grpc.start_time": "2023-03-30T15:48:05-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "StreamStatus", "grpc.code": "OK", "grpc.time_ms": 501.601}
2023-03-30T15:48:06.044-0400    WARN    robot_server    zap/options.go:212      finished client unary call      {"system": "grpc", "span.kind": "client", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetStatus", "error": "rpc error: code = Unimplemented desc = ", "grpc.code": "Unimplemented", "grpc.time_ms": 0.98}
2023-03-30T15:48:06.045-0400    INFO    robot_server    zap/options.go:212      finished streaming call with code OK    {"grpc.start_time": "2023-03-30T15:48:05-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "StreamStatus", "grpc.code": "OK", "grpc.time_ms": 501.982}
2023-03-30T15:48:06.077-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "SendSessionHeartbeat", "grpc.code": "OK", "grpc.time_ms": 0.177}
2023-03-30T15:48:06.266-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "SendSessionHeartbeat", "grpc.code": "OK", "grpc.time_ms": 0.156}
2023-03-30T15:48:06.501-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "ResourceNames", "grpc.code": "OK", "grpc.time_ms": 0.17}
2023-03-30T15:48:06.502-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "ResourceNames", "grpc.code": "OK", "grpc.time_ms": 0.156}
2023-03-30T15:48:06.503-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "ResourceNames", "grpc.code": "OK", "grpc.time_ms": 0.186}
2023-03-30T15:48:06.508-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "SendSessionHeartbeat", "grpc.code": "OK", "grpc.time_ms": 0.154}
2023-03-30T15:48:06.539-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetOperations", "grpc.code": "OK", "grpc.time_ms": 0.18}
2023-03-30T15:48:06.539-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetOperations", "grpc.code": "OK", "grpc.time_ms": 0.091}
2023-03-30T15:48:06.569-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetSessions", "grpc.code": "OK", "grpc.time_ms": 0.199}
2023-03-30T15:48:06.570-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetSessions", "grpc.code": "OK", "grpc.time_ms": 0.188}
2023-03-30T15:48:06.711-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "SendSessionHeartbeat", "grpc.code": "OK", "grpc.time_ms": 0.15}
2023-03-30T15:48:06.727-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetOperations", "grpc.code": "OK", "grpc.time_ms": 0.453}
2023-03-30T15:48:06.760-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetSessions", "grpc.code": "OK", "grpc.time_ms": 0.338}
2023-03-30T15:48:06.938-0400    INFO    robot_server    zap/options.go:212      finished unary call with code OK        {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "SendSessionHeartbeat", "grpc.code": "OK", "grpc.time_ms": 0.205}
2023-03-30T15:48:07.041-0400    WARN    robot_server    zap/options.go:212      finished client unary call      {"system": "grpc", "span.kind": "client", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetStatus", "error": "rpc error: code = Unimplemented desc = ", "grpc.code": "Unimplemented", "grpc.time_ms": 0.898}
2023-03-30T15:48:07.042-0400    INFO    robot_server    zap/options.go:212      finished streaming call with code OK    {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "StreamStatus", "grpc.code": "OK", "grpc.time_ms": 504.461}
2023-03-30T15:48:07.042-0400    WARN    robot_server    zap/options.go:212      finished client unary call      {"system": "grpc", "span.kind": "client", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "GetStatus", "error": "rpc error: code = Unimplemented desc = ", "grpc.code": "Unimplemented", "grpc.time_ms": 2.175}
2023-03-30T15:48:07.043-0400    INFO    robot_server    zap/options.go:212      finished streaming call with code OK    {"grpc.start_time": "2023-03-30T15:48:06-04:00", "system": "grpc", "span.kind": "server", "grpc.service": "viam.robot.v1.RobotService", "grpc.method": "StreamStatus", "grpc.code": "OK", "grpc.time_ms": 504.272}
^C2023-03-30T15:48:07.048-0400  INFO    robot_server.process.intel_/usr/local/bin/intelrealgrpcserver   pexec/managed_process_unix.go:71        stopping process 5850 with signal terminated
```

**Pull request merged**